clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -564,6 +564,51 @@ class radix_sort {
   uint8_t *_local_memory;
 };
 
+/// Load linear segment items into block format across threads
+/// Helper for Block Load
+template <int GROUP_THREADS, typename InputT, int ITEMS_PER_THREAD,
+          typename InputIteratorT>
+__dpct_inline__ void load_blocked(int linear_tid, InputIteratorT block_itr,
+                                  InputT (&items)[ITEMS_PER_THREAD]) {
+#pragma unroll
+  for (int ITEM = 0; ITEM < ITEMS_PER_THREAD; ITEM++) {
+
+    if ((linear_tid * ITEMS_PER_THREAD) + ITEM < GROUP_THREADS) {
+
+      items[ITEM] = block_itr[(linear_tid * ITEMS_PER_THREAD) + ITEM];
+    }
+  }
+}
+
+template <int GROUP_THREADS, typename InputT, int ITEMS_PER_THREAD,
+          typename InputIteratorT>
+__dpct_inline__ void load_striped(int linear_tid, InputIteratorT block_itr,
+                                  InputT (&items)[ITEMS_PER_THREAD]) {
+#pragma unroll
+  for (int ITEM = 0; ITEM < ITEMS_PER_THREAD; ITEM++) {
+
+    if((linear_tid + (ITEM * ITEMS_PER_THREAD) < GROUP_THREADS){
```

Comment:
Need reformat and I think the if-stmt can be removed.

Reply:
The if-stmt is put to guard against GROUP_THREAD bounds (as a range) . reformatted.

Reply:
I believe @yihanwg  means that we may want to rethink if statement **inside** the inner loop to avoid overflow for `GROUP_WORK_ITEMS`.  We should be able to do some math outside the look and then have special casing for "full" and "partial" loads or just simply have an upper bound of the loop controlled by the starting math.  Checking this at every iteration seems very slow, and we should be able to know how many iterations this work_item needs to run with a single check or min / max operation.

We may need to consider thread divergence here, but I think even with that in mind this wont be the optimal way of writing this.

Reply:
I see that the if statements have been removed, but they have not been replaced by anything.  I assume that we still ultimately need to protect overrunning the range. Also, it is important for unrolling that the loop bounds are known at compile time.  Therefore, we may look at a sort of branched approach.

Here is some pseudocode:

``` 
if  (last element will overrun range)
{
   upper_bound = calculate last element which fits in range
   for (loop from start to upper_bound without unrolling)
       operation //no if statement guard is required
}
else
{
   #pragma unroll
   for (loop from start to end with no checks)
       operation //no if statement guard is required
}
```